ni_measurementlink_service/_internal/parameter/serializer.py

```diff
@@ -216,22 +215,25 @@ def _deserialize_enum_parameters(
         None: No return value.
 
     """
-    for i, value in parameter_by_id.items():
-        parameter_metadata = parameter_metadata_dict[i]
-        has_enum_values_annotation, enum_values_annotation = try_get_enum_values_annotation(
-            parameter_metadata
-        )
-        if has_enum_values_annotation:
+    for id, value in parameter_by_id.items():
+        parameter_metadata = parameter_metadata_dict[id]
+        if get_enum_values_annotation(parameter_metadata):
             enum_type = _get_enum_type(parameter_metadata)
             if parameter_metadata.repeated:
-                for j, member_value in enumerate(value):
-                    for enum in enum_type:
-                        if enum.value == member_value:
-                            parameter_by_id[i][j] = enum
+                for index, member_value in enumerate(value):
+                    try:
+                        parameter_by_id[id][index] = enum_type(member_value)
+                    except ValueError:
+                        raise TypeError(
+                            f"Unexpected value {member_value} is not valid for enum type '{enum_type.__name__}'."
+                        )
```

Comment:
https://docs.python.org/3/library/exceptions.html#TypeError

> Passing arguments of the wrong type (e.g. passing a [list](https://docs.python.org/3/library/stdtypes.html#list) when an [int](https://docs.python.org/3/library/functions.html#int) is expected) should result in a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError), but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError).

Seems like "unexpected value X" should be a ValueError. 

Do you really need to customize the error message? The default is something like `ValueError: 1234 is not a valid UsageTypeAI`.

Reply:
Removed custom error message. I agree that default should be fine.